[libteam] Support multiple member ports for LACP fallback

lukasstockner opened the pull request:
#### Why I did it
Fixes #7555 and #8866.

While the LACP fallback HLD correctly describes the behavior for LAGs with multiple ports, the actual implementation only worked for single-port LAGs.

When multiple ports are present, the fallback mode would select all ports, therefore making it useless for the purpose described in the HLD (PXE-booting).

#### How I did it
This commit adds a patch to implement the desired behavior:
If all non-disabled ports are defaulted, *one* port goes into fallback mode and gets selected for the LAG. As soon as any port receives a LACPDU, the fallback port goes back to defaulted.
The fallback port is selected by higher port priority or, if both have equal priority (the default), by lower port number (as described in the HLD).

#### How to verify it
See #7555 - start a DVS container, create a Portchannel with two members and fallback enabled, create a bond on the non-DVS end of the interfaces, bring it up and down and observe the state of the Portchannel.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [x] 202106
- [x] 202111

Not sure if this counts as a fix or feature, but I'd argue fix since this behavior was described in the HLD all along and the fallback is not particularly useful without it.

#### Description for the changelog
Support multiple member ports for LACP fallback.

#### Link to config_db schema for YANG module changes
#### A picture of a cute animal (not mandatory but encouraged)
![IMG-1132](https://user-images.githubusercontent.com/10601319/168190017-1f5d97db-555c-4d69-8e85-1dc0ce811a88.jpg)

lukasstockner commented:
@saiarcot895 @vaibhavhd based on recent PRs, you might be the right people to ask for a review here? Any help would be appreciated.

puffc commented:
@rodnymolina @liat-grozovik Could you please review this PR?

puffc commented:
@saiarcot895 Ping

saiarcot895 commented on src/libteam/patch/0015-Extend-LACP-fallback-support-to-multiple-ports.patch:
```diff
@@ -0,0 +1,253 @@
++		}
++		if (lacp_port->state == PORT_STATE_CURRENT || lacp_port->state == PORT_STATE_EXPIRED) {
++			/* If at least one port is currently not defaulted, don't do any fallback. */
++			have_selected_port = true;
```
Could you rename this to something like `have_active_port` or similar? `have_selected_port` sounds like a port was selected in this loop.

lukasstockner replied:
I'm not certain that `selected` is perfectly accurate here, so I've renamed this to `do_fallback`, which is a better description of the purpose anyways - it's supposed to track whether we want to use fallback logic or not.

saiarcot895 commented:
Note that you'll need to resolve the merge conflicts as well.

puffc commented on src/libteam/patch/series:
```diff
@@ -12,3 +12,4 @@
 0012-Increase-min_ports-upper-limit-to-1024.patch
 0013-set-port-to-disabled-state-during-removal.patch
 0014-dont-move-the-port-state-from-disabled-when-admin-state-is-down.patch
+0015-Extend-LACP-fallback-support-to-multiple-ports.patch
```
@lukasstockner Ping

lukasstockner replied:
Rebased now, sorry for the delay.

lukasstockner committed:
[libteam] Support multiple ports with LACP fallback

Fixes #7555 and #8866.

While the LACP fallback HLD correctly describes the behavior for LAGs with
multiple ports, the actual implementation only worked for single-port LAGs.

When multiple ports are present, the fallback mode would select all ports,
therefore making it useless for the purpose described in the HLD (PXE-booting).

Therefore, this commit adds a patch to implement the desired behavior:
If all non-disabled ports are defaulted, *one* port goes into fallback mode and
gets selected for the LAG. As soon as any port receives a LACPDU, the fallback
port goes back to defaulted.
The fallback port is selected by higher port priority or, if both have equal
priority (the default), by lower port number (as described in the HLD).

Signed-off-by: Lukas Stockner <lstockner@genesiscloud.com>

puffc commented:
@saiarcot895 Please review this PR again. Thanks.

saiarcot895 commented:
/azpw run Azure.sonic-buildimage

yxieca reviewed (commented):
### Review summary

Thanks for this fix — and apologies it sat so long. I did a thorough pass (logic trace + protocol behavior + the libteam patch-stack interactions). **The core change is correct and conforms to the LACP Fallback HLD**: it replaces the old per-port `is_lacp_fallback_eligible()` (which made *every* defaulted port eligible → all members selected, the bug in #7555/#8866) with an explicit single-port election via a new `PORT_STATE_FALLBACK`. I verified:

- **State-machine substitutions are 1:1 and strictly narrower** — every former `is_lacp_fallback_eligible(p)` site (`selectable_state`, `unselectable_state`, `should_be_enabled`, `should_be_disabled`, `actor_update`'s `INFO_STATE_DEFAULTED`) now keys on `state == PORT_STATE_FALLBACK`, so only the one elected port is selected/enabled without partner sync. ✔
- **Election is deterministic and HLD-conformant** — `port_priority` (higher wins) then port number (lower wins) matches the HLD tie-break. ✔
- **Re-entrancy terminates** — `lacp_port_set_state → lacp_ports_update_fallback_state → lacp_port_set_state(old→DEFAULTED) → … → (new→FALLBACK)` converges in ~3 passes because the election is stable on unchanging actor fields (as your inline comment notes). ✔
- **Teardown on partner appearance** — any member reaching CURRENT/EXPIRED sets `do_fallback=false`, demoting the fallback port back to DEFAULTED. ✔
- **`teamdctl.c` JSON widening is safe** — the daemon already emits `runner.fallback` (added by patch `0003-libteam-Add-fallback-support-for-single-member-port-`), so the extended `json_unpack("{s:{s:b, s:i, s:b, s:b}}")` parses and `state view` won't break. ✔

### Blocking before merge

**1. Patch-number collision (this is why the PR is CONFLICTING).** Master's `src/libteam/patch/series` now already ends with its own `0017-Specify-netlink-recv-buffer-size-to-avoid-a-syscall.patch` and `0018-Replace-FD_SET-select-with-poll-to-avoid-FD_SETSIZE-abort.patch` (both added since this PR was opened). Please **renumber this patch to `0019-...`**, update the `series` append accordingly, and re-verify the hunks still apply cleanly against the current libteam **1.31** patched tree (the runner-logic hunks depend on 0001–0016, esp. 0003, being applied first).

### Minor (non-blocking)

**2. Misleading "lower system wins" comment in `lacp_port_better_for_fallback`.** The first comparison is `memcmp(actor1->system, actor2->system, ETH_ALEN)`, but `actor.system` is set from `ctx->hwaddr` (the LAG-wide system MAC) — so within a single LAG it's identical across all members and `system_diff` is effectively always `0`; the real selectors are `port_priority` then port number. The comparison is harmless (and mirrors upstream's `lacp_port_better()` comparator style), but a one-line comment noting that the system-MAC key only matters across differing systems would save a future reader some confusion.

No automated test ships with the patch (expected for a vendored libteam patch — verification is the manual DVS repro in #7555); the protocol logic itself looks complete and correct.
